greedybear/cronjobs/extraction/strategies/cowrie.py

```
@@ -106,50 +122,95 @@ def _get_url_downloads(self, hits: list[dict]) -> None:
-            for hit in sorted(hits, key=lambda hit: hit["timestamp"]):
-                match hit["eventid"]:
-                    case "cowrie.session.connect":
-                        session_record.start_time = hit["timestamp"]
-                    case "cowrie.login.failed" | "cowrie.login.success":
-                        session_record.login_attempt = True
-                        username = hit["username"].replace("\x00", "[NUL]")
-                        password = hit["password"].replace("\x00", "[NUL]")
-                        session_record.credentials.append(f"{username} | {password}")
-                        session_record.source.login_attempts += 1
-                    case "cowrie.command.input":
-                        self.log.info(f"found a command execution from {ioc.name}")
-                        session_record.command_execution = True
-                        if session_record.commands is None:
-                            session_record.commands = CommandSequence()
-                            session_record.commands.first_seen = hit["timestamp"]
-                        command = hit["message"].removeprefix("CMD: ").replace("\x00", "[NUL]")
-                        session_record.commands.last_seen = hit["timestamp"]
-                        session_record.commands.commands.append(command[:1024])
-                    case "cowrie.session.closed":
-                        session_record.duration = hit["duration"]
-                session_record.interaction_count += 1
+    def _save_sessions(self, sessions: Dict[str, CowrieSessionData]) -> None:
+        for sid, session_data in sessions.items():
+            source_ip = session_data.source_ip
+            # We need the IOC object for the source IP
+            # It should have been created by _get_scanners
+            scanner_ioc = self.ioc_repo.get_ioc_by_name(source_ip)
+            if not scanner_ioc:
+                # Should not happen if data consistency holds, but fallback:
```

I think I would throw an exception or at least log an error and continue. Because at this point _get_scanners() already finished, and if there is no IOC something in the code went horribly wrong, right?